feat: packaging

LostNeophyte opened the pull request:
adds the cmake setup from LunarVim-mono.
the ci job will create nightly releases on a schedule and create stable releases on tag push, if the stable release already exists it will upload files to it.

you can see it in action here
https://github.com/LostNeophyte/LunarVim/releases
https://github.com/LostNeophyte/LunarVim/actions/runs/5435290043

kylo252 reviewed (changes requested):
why not invoke the cmake script from that repo instead? basically clone it first and then run whatever build steps you want.
You can even create a custom github job from that repo to get a cleaner interface.

pros:
- avoid adding a lot of files that are only relevant for distribution
- won't clutter the repo's history with `ci` commits
- adjusting build steps faster and easier

LostNeophyte commented:
cmake needs to be together with the source code for homebrew

osalbahr commented:
> why not invoke the cmake script from that repo instead? basically clone it first and then run whatever build steps you want. You can even create a custom github job from that repo to get a cleaner interface.
> 
> pros:
> 
> * avoid adding a lot of files that are only relevant for distribution
> * won't clutter the repo's history with `ci` commits
> * adjusting build steps faster and easier

Cloning would make the build artifacts non-deterministic and Homebrew compares against a shasum. See Homebrew/homebrew-core#132446 for more context.

kylo252 commented:
> Cloning would make the build artifacts non-deterministic and Homebrew compares against a shasum. See [Homebrew/homebrew-core#132446](https://github.com/Homebrew/homebrew-core/pull/132446) for more context. Specifically [this comment](https://github.com/Homebrew/homebrew-core/pull/132446/files#r1212718069).

@osalbahr, thanks for adding more context!

I'm still not sure about adding all those files into the repo, I was hoping we can get away with doing something along the these lines

```ruby
class Lunarvim < Formula
  desc "IDE layer for Neovim. Completely free and community driven"
  homepage "https://www.lunarvim.org"
  url "https://github.com/LunarVim/LunarVim/archive/refs/tags/1.3.0.tar.gz"
  version "1.3.0"
  sha256 "2b6cefdf76d42e2d349381e63d42fa3c3e2527da25a464cb0a2af47bdea9e2a0"
  license "GPL-3.0-or-later"
  head "https://github.com/LunarVim/LunarVim.git", branch: "master"

  resource "LunarVim-mono" do
    url "https://github.com/LunarVim/LunarVim-mono/archive/refs/tags/stable.tar.gz" # replace stable with 1.3.0
    sha256 "4b63e968655d9a71ff32fdca653a56bd560555c1780d704669df578048b5f808"
  end

  depends_on "cmake" => :build
  depends_on "fd"
  depends_on "neovim"
  depends_on "node"
  depends_on "python@3.11"
  depends_on "ripgrep"
  depends_on "tree-sitter"

  def install
    resources.each do |r|
      r.stage(buildpath/"packaging"/r.name)
    end
    cd "packaging" do
      # ..
    end
  end

  test do
    assert shell_output("#{bin}/lvim -v").start_with?("NVIM")
    assert_equal "lmao", pipe_output("#{bin}/lvim -Es +%print", "lmao\n").strip
  end
end
```

@ChristianChiarulli, what's your stance on this?

LostNeophyte commented:
> why not invoke the cmake script from that repo instead? basically clone it first and then run whatever build steps you want. You can even create a custom github job from that repo to get a cleaner interface.
> 
> pros:
> 
>     * avoid adding a lot of files that are only relevant for distribution
> 
>     * won't clutter the repo's history with `ci` commits
> 
>     * adjusting build steps faster and easier

I could probably make this pr quite a bit smaller. the build step would stay here, that would leave 151 loc (7 files 2 of which are icons) in this pr. the rest (github workflow, appimage, Packaging.cmake, BundlePlugins.cmake) would go to the LunarVim-mono repo.

kylo252 commented:
> I could probably make this pr quite a bit smaller. the build step would stay here, that would leave 151 loc (7 files 2 of which are icons) in this pr. the rest (github workflow, appimage, Packaging.cmake, BundlePlugins.cmake) would go to the LunarVim-mono repo.

Why not keep all "lunarvim-packaging" scripts/tools in a separate repo? You'd clone it into `./utils/packaging` and invoke the cmake command there (both of these steps can be done through the main `Makefile`)

Keep in mind we probably need more scripts for integrating with other package formats (e.g. nix, scoop, winget,. etc.) so this packaging repo will only get more complex.

LostNeophyte commented:
> > I could probably make this pr quite a bit smaller. the build step would stay here, that would leave 151 loc (7 files 2 of which are icons) in this pr. the rest (github workflow, appimage, Packaging.cmake, BundlePlugins.cmake) would go to the LunarVim-mono repo.
> 
> Why not keep all "lunarvim-packaging" scripts/tools in a separate repo? You'd clone it into `./utils/packaging` and invoke the cmake command there (both of these steps can be done through the main `Makefile`)
> 
> Keep in mind we probably need more scripts for integrating with other package formats (e.g. nix, scoop, winget,. etc.) so this packaging repo will only get more complex.

I can agree with all the packaging stuff begin in a separate repo. just the build cmake would stay in the core, it makes packaging for others way easier and it makes sense for it to be versioned. I think it would be a nice compromise

kylo252 commented:
> I can agree with all the packaging stuff begin in a separate repo. just the build cmake would stay in the core, it makes packaging for others way easier and it makes sense for it to be versioned. I think it would be a nice compromise

Not sure which cmake files you mean here, since the main `CMakeLists.txt` wouldn't contain much logic anymore, and the cpack recipe is what I'm thinking can go in a separate repo.

Although, looking more into it, I see some ways to minimize the added files. 

Please try moving the cmake logic into just two files: `CMakeLists.txt` and a cpack recipe. Let me know if you need help with that.

LostNeophyte commented:
so now these files would be in the core:
- CMakeLists.txt
- utils/icons/*
- utils/packaging/assets/lvim & - utils/packaging/assets/lvim.ps1 # these could be merged into the current bin templates and the cmake script would just remove the -u flag from the exec in the bash script

the rest would go to the lunarvim-mono repo

kylo252 reviewed (changes requested):
-  Let's merge the icon files into `utils/desktop`
- `CMakeLists.txt` should not be on the top level, it makes it look like it's necessary to use it to "build" the repo/config (I think should be possible to do)

LostNeophyte commented:
> CMakeLists.txt should not be on the top level, it makes it look like it's necessary to use it to "build" the repo/config (I think should be possible to do)

I tried to do that, and it seems like a pain because you'd need to access paths outside of the cmake project, but i'll do a bit more research on it

kylo252 commented:
> I tried to do that, and it seems like a pain because you'd need to access paths outside of the cmake project, but i'll do a bit more research on it

One approach might be to use `CPackConfig.cmake` instead of a `CMakeLists.txt`, then you can call that directly, see https://gitlab.kitware.com/cmake/community/-/wikis/doc/cpack/Packaging-With-CPack

LostNeophyte commented:
my goal here is just to have something that will simplify packaging for multiple formats so that we won't have to do the same steps for each package, I think just CPack without cmake won't help with that, we could modify our current installer to allow this usage (or create a new simple one, in which case the current installer could use the simpler one). it would need these modifications, maybe more:
- allow skipping all the extra steps
- allow using a local repo and copy it instead of linking
- and probably allow skipping the bootstrapping and running tests, some package repos probably won't be ok with downloading plugins, but it would be nice to have that for our own packages to allow offline install and usage

what do you think about this?

osalbahr commented:
I would highly appreciate if installing/upgrading lunarvim was just one command!

LostNeophyte commented:
> I would highly appreciate if installing/upgrading lunarvim was just one command `brew install lunarvim`! Currently I have to install `neovim` with Homebrew and then run the LunarVim script since neovim it’s outdated in the repos of many distros.

yea, this is my goal with this, the current way of installing and updating lvim is unusual, unintuitive and kinda annoying

we won't really be able to solve the outdated neovim thing, we only have control over that in the appimage where neovim would be bundled with lunarvim like it is in [these releases](https://github.com/LunarVim/LunarVim-mono/releases)

osalbahr commented:
Is there a reason why the install artifacts can’t be on the main repo? Homebrew basically just automates running the manual install instructions with the added benefit of also handling upgrades and dependencies.

LostNeophyte commented:
I personally would like them to be in this repo, but Kylo doesn't and I can live with them being in a separate one so the artifacts will probably be in LunarVim/LunarVim-mono like they are now, there aren't really that many downsides to that

osalbahr commented:
> I personally would like them to be in this repo, but Kylo doesn't and I can live with them being in a separate one so the artifacts will probably be in LunarVim/LunarVim-mono like they are now, there aren't really that many downsides to that

I’d ask in Homebrew if this is acceptable. This is strange to me.

osalbahr commented:
> I personally would like them to be in this repo, but Kylo doesn't and I can live with them being in a separate one so the artifacts will probably be in LunarVim/LunarVim-mono like they are now, there aren't really that many downsides to that

I’d ask in Homebrew if this is acceptable. This is strange to me. Aren’t the install artifacts all in the source code? How would one install LunarVim from source?

LostNeophyte commented:
in this repo there would be an installer/cmake script/something else that would be able to install the lunarvim version from the local folder that the installer is in
that installer would be used in the Lunarvim-mono repo to create an appimage, zip and a few different things
and that same installer would be used in homebrew

osalbahr commented:
> in this repo there would be an installer/cmake script/something else that would be able to install the lunarvim version from the local folder that the installer is in
> that installer would be used in the Lunarvim-mono repo to create an appimage, zip and a few different things
> and that same installer would be used in homebrew and all other places to create packages

LunarVim-mono, public archive repo, is the official way to install LunarVim from source?

I really like this project but I don’t know why such a popular project is only available in the AUR. I’d like to help with packaging to mainstream sources if needed.

https://repology.org/project/lunarvim/versions

osalbahr commented:
> yea, we currently have no real way to install lvim from source, so packaging lvim is mostly impossible and that's what I want to change

I think this is a great feature that would benefit Homebrew and non-Homebrew users alike!

> the aur package is broken since 1.2 and no longer maintained I believe (it's not official, I don't think the author of the package reached out to us)

Oh, makes sense. I actually don’t use the AUR very often so I don’t know how it works.

kylo252 commented:
@LostNeophyte, I actually figured out a pretty simple solution that will, hopefully, satisfy everyone!

1. move the cmake files into `utils/packaging`
2. add appropriate `make` rules that will copy `CMakeLists.txt` and then invoke the correct cmake command, e.g. `make package-brew`

This will avoid any confusion about having a `CMakeLists.txt` in the repo's root when it's not really used for building (since this isn't a C/C++ project), basically solving my main complaint.

LostNeophyte commented:
> @LostNeophyte, I actually figured out a pretty simple solution that will, hopefully, satisfy everyone!
> 
>     1. move the cmake files into `utils/packaging`
> 
>     2. add appropriate `make` rules that will copy `CMakeLists.txt` and then invoke the correct cmake command, e.g. `make package-brew`
> 
> 
> This will avoid any confusion about having a `CMakeLists.txt` in the repo's root when it's not really used for building (since this isn't a C/C++ project), basically solving my main complaint.

Sounds good to me, do I still put the ci workflow, appimage script and CPack in the lvim-mono repo?

osalbahr commented:
> when it's not really used for building (since this isn't a C/C++ project)

what is used for building LunarVim from source?

LostNeophyte commented:
> what is used for building LunarVim from source?

we don't build it, lunarvim doesn't require building. the installer only moves files to correct places. well, maybe you could argue that the `lvim` shell script is built, in that case we use `sed` to build it (just substituting a couple strings for LUNARVM_*_DIR env variables)

osalbahr commented:
> we don't build it, lunarvim doesn't require building. the installer only moves files to correct places. well, maybe you could argue that the `lvim` shell script is built, in that case we use `sed` to build it (just substituting a couple strings for LUNARVM_*_DIR env variables)

That’s great. If you list the exact step-by-step commands, we can just put these commands in the Homebrew formula. It seems no changes to the repo would be needed. Here is an simple example of a “def install” they can be just shell commands, but it would be easier to maintain if the commands were wrapped in a “./configure --PREFIX=<insert Homebrew prefix>” and then “make install” although not required

The added benefit would be that someone manually installing from the master branch can also follow the same steps in the Homebrew formula. You can think of the “def install” section as a general and reproducible way of manually installing from source https://docs.brew.sh/Formula-Cookbook#grab-the-url.

I think maybe that the script in the LunarVim basically does that? If so then we can probably use it.

LostNeophyte commented:
> > we don't build it, lunarvim doesn't require building. the installer only moves files to correct places. well, maybe you could argue that the `lvim` shell script is built, in that case we use `sed` to build it (just substituting a couple strings for LUNARVM_*_DIR env variables)
> 
> That’s great. If you list the exact step-by-step commands, we can just put these commands in the Homebrew formula. It seems no changes to the repo would be needed. Here is an simple example of a “def install” they can be just shell commands: https://docs.brew.sh/Formula-Cookbook#grab-the-url
>
> but it would be easier to maintain if the commands were wrapped in a “./configure --PREFIX=” and then “make install” although not required.
>
> The added benefit would be that someone manually installing from the master branch can also follow the same steps in the Homebrew formula. You can think of the “def install” section as a general and reproducible way of manually installing from source.

yup, I don't want the commands to be in the formula, as they'd need to be replicated in many places.
I'll use cmake instead of ./configure because of windows, but since CMakeLists.txt will be in a subfolder it will be impossible to use cmake normally so I'll add a new make install target to move CMakeLists.txt and run cmake (with DESTDIR support) or a make configure target that will only move CMakeLists.txt to the root directory, I'm not sure which option makes more sense, or maybe there's a better option I'm missing 

> I think maybe that the script in the LunarVim basically does that? If so then we can probably use it.

it's close, not quite there. I think it's complex enough and I'd rather not add a new feature to it

osalbahr commented:
> yup, I don't want the commands to be in the formula, as they'd need to be replicated in many places.

Great point. I agree.

> I'll use cmake instead of ./configure because of windows

Interesting. I haven’t done much development on windows this doesn’t surprise. 

> but since CMakeLists.txt will be in a subfolder it will be impossible to use cmake normally so I'll add a new make install target to move CMakeLists.txt

That’s a nice workaround! Not sure if doing `cd subfolder` has any disadvantages, not really familiar with the cmake project structure. It seems it would simplify things as to not have to `../` everything.

kylo252 commented:
> Sounds good to me, do I still put the ci workflow, appimage script and CPack in the lvim-mono repo?

no, that's fine since it goes into `.github`. 

nowever, we should look into making it a github action if the scripts end up being too long/complex.

> I'll add a new make install target to move CMakeLists.txt and run cmake (with DESTDIR support) or a make configure target that will only move CMakeLists.txt to the root directory, I'm not sure which option makes more sense, or maybe there's a better option I'm missing

it's something like this

```diff
diff --git a/Makefile b/Makefile
index 1f6844c5..bc842db5 100644
--- a/Makefile
+++ b/Makefile
@@ -39,4 +39,9 @@ style-sh:
 test:
 	bash ./utils/ci/run_test.sh "$(TEST)"
 
-.PHONY: install install-neovim-binary uninstall lint style test
+packaging-brew:
+	cp utils/packaging/CMakeLists.txt.in CMakeLists.txt
+	cmake --configure . -B build -D BUNDLE_PLUGINS=1
+	cpack
+
+.PHONY: packaging-brew install install-neovim-binary uninstall lint style test
```

kylo252 commented:
> Sounds good to me, do I still put the ci workflow, appimage script and CPack in the lvim-mono repo?

no, that's fine since it goes into `.github/`. 

however, we should look into making it a github action if the scripts end up being too long/complex.

> I'll add a new make install target to move CMakeLists.txt and run cmake (with DESTDIR support) or a make configure target that will only move CMakeLists.txt to the root directory, I'm not sure which option makes more sense, or maybe there's a better option I'm missing

it's something like this

```diff
diff --git a/Makefile b/Makefile
index 1f6844c5..bc842db5 100644
--- a/Makefile
+++ b/Makefile
@@ -39,4 +39,9 @@ style-sh:
 test:
 	bash ./utils/ci/run_test.sh "$(TEST)"
 
-.PHONY: install install-neovim-binary uninstall lint style test
+packaging-brew:
+	cp utils/packaging/CMakeLists.txt.in CMakeLists.txt
+	cmake --configure . -B build -D BUNDLE_PLUGINS=1
+	cpack
+
+.PHONY: packaging-brew install install-neovim-binary uninstall lint style test
```

chenrui333 commented:
looks like homebrew PR is stale, might need another one to bring life to it.

LostNeophyte commented:
@kylo252 
I moved CMakeLists.txt out of the root,
moved the icons to the packaging folder,
added a make target `configure` that links CMakeLists to the root, this allows passing flags to cmake

kylo252 commented on utils/packaging/CMakeLists.txt.in:
```diff
@@ -0,0 +1,113 @@
+    DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/icons/hicolor/256x256/apps)
+endif()
+
+# bake the lvim version into the code  TODO: find a better way to do this
```
you can tell CMake to generate a file based on a template, it's pretty straightforward and allows you to pass variables, see:
- https://github.com/neovim/neovim/blob/81d5550d77d24a924e8c23873fccbf405bc39447/cmake.config/versiondef.h.in
- https://cmake.org/cmake/help/latest/command/configure_file.html

kylo252 commented on utils/packaging/assets/lvim:
we should re-use the files under `utils/bin` or change the installer(s) to use these ones instead.

you can make use of `configure_file()` here as well and tweak the placeholder variable to be compatible with both CMake and bash, it would look something like this:

```diff
diff --git a/utils/bin/lvim.template b/utils/bin/lvim.template
index 4c04b9f8..af2943b5 100644
--- a/utils/bin/lvim.template
+++ b/utils/bin/lvim.template
@@ -1,11 +1,11 @@
 #!/usr/bin/env bash
 
-export NVIM_APPNAME="${NVIM_APPNAME:-NVIM_APPNAME_VAR}"
+export NVIM_APPNAME="${NVIM_APPNAME:-@NVIM_APPNAME_VAR@}"
 
-export LUNARVIM_RUNTIME_DIR="${LUNARVIM_RUNTIME_DIR:-RUNTIME_DIR_VAR}"
-export LUNARVIM_CONFIG_DIR="${LUNARVIM_CONFIG_DIR:-CONFIG_DIR_VAR}"
-export LUNARVIM_CACHE_DIR="${LUNARVIM_CACHE_DIR:-CACHE_DIR_VAR}"
+export LUNARVIM_RUNTIME_DIR="${LUNARVIM_RUNTIME_DIR:-@RUNTIME_DIR_VAR@}"
+export LUNARVIM_CONFIG_DIR="${LUNARVIM_CONFIG_DIR:-@CONFIG_DIR_VAR@}"
+export LUNARVIM_CACHE_DIR="${LUNARVIM_CACHE_DIR:-@CACHE_DIR_VAR@}"
 
-export LUNARVIM_BASE_DIR="${LUNARVIM_BASE_DIR:-BASE_DIR_VAR}"
+export LUNARVIM_BASE_DIR="${LUNARVIM_BASE_DIR:-@BASE_DIR_VAR@}"
 
 exec -a "$NVIM_APPNAME" nvim -u "$LUNARVIM_BASE_DIR/init.lua" "$@"
diff --git a/utils/installer/install_bin.sh b/utils/installer/install_bin.sh
index a25ad370..15ef0e39 100755
--- a/utils/installer/install_bin.sh
+++ b/utils/installer/install_bin.sh
@@ -25,11 +25,11 @@ function setup_shim() {
 
   cp "$src" "$dst"
 
-  sed -e s"#NVIM_APPNAME_VAR#\"${NVIM_APPNAME}\"#"g \
-    -e s"#RUNTIME_DIR_VAR#\"${LUNARVIM_RUNTIME_DIR}\"#"g \
-    -e s"#CONFIG_DIR_VAR#\"${LUNARVIM_CONFIG_DIR}\"#"g \
-    -e s"#CACHE_DIR_VAR#\"${LUNARVIM_CACHE_DIR}\"#"g \
-    -e s"#BASE_DIR_VAR#\"${LUNARVIM_BASE_DIR}\"#"g "$src" \
+  sed -e s"#@NVIM_APPNAME_VAR@#\"${NVIM_APPNAME}\"#"g \
+    -e s"#@RUNTIME_DIR_VAR@#\"${LUNARVIM_RUNTIME_DIR}\"#"g \
+    -e s"#@CONFIG_DIR_VAR@#\"${LUNARVIM_CONFIG_DIR}\"#"g \
+    -e s"#@CACHE_DIR_VAR@#\"${LUNARVIM_CACHE_DIR}\"#"g \
+    -e s"#@BASE_DIR_VAR@#\"${LUNARVIM_BASE_DIR}\"#"g "$src" \
     | tee "$dst" >/dev/null
 
   chmod u+x "$dst"
```

kylo252 commented on utils/packaging/cmake/BundlePlugins.cmake:
```diff
@@ -0,0 +1,55 @@
+set(INIT_LUA_PATH ${LUNARVIM_BASE_DIR}/init.lua)
+  
+if(WIN32)
+  string(REPLACE "/" "\\" INIT_LUA_PATH "${GEN_PLUGINS_INIT_PATH}")
+  string(REPLACE "/" "\\" LVIM_XDG_DATA_HOME "${LVIM_XDG_DATA_HOME}")
+  string(REPLACE "/" "\\" LVIM_XDG_CONFIG_HOME "${LVIM_XDG_CONFIG_HOME}")
+  string(REPLACE "/" "\\" LVIM_XDG_CACHE_HOME "${LVIM_XDG_CACHE_HOME}")
+  string(REPLACE "/" "\\" LUNARVIM_RUNTIME_DIR "${LUNARVIM_RUNTIME_DIR}")
+  string(REPLACE "/" "\\" LUNARVIM_CONFIG_DIR "${LUNARVIM_CONFIG_DIR}")
+  string(REPLACE "/" "\\" LUNARVIM_CACHE_DIR "${LUNARVIM_CACHE_DIR}")
+  string(REPLACE "/" "\\" LUNARVIM_BASE_DIR "${LUNARVIM_BASE_DIR}")
```
is this really necessary?

kylo252 commented on utils/packaging/cmake/BundlePlugins.cmake:
```diff
@@ -0,0 +1,55 @@
+set(XDG_ROOT ${CMAKE_BINARY_DIR}/xdg_root)
+set(LVIM_XDG_DATA_HOME ${XDG_ROOT}/share)
+set(LVIM_XDG_CONFIG_HOME ${XDG_ROOT}/.config)
+set(LVIM_XDG_CACHE_HOME ${XDG_ROOT}/.cache)
```
I don't have the full picture but I feel like this would require moving the plugins again since `xdg_root` is hardcoded in the path. 

This should be easy to verify though, so please show the intended folder structure.

kylo252 commented on utils/packaging/cmake/BundlePlugins.cmake:
```diff
@@ -0,0 +1,55 @@
+message("downloading plugins...")
+message("${DOWNLOAD_PLUGINS_CMD}")
+
+set(BUNDLE_PLUGINS_TIMEOUT 120 CACHE STRING "")
```
I think this limit might be too short on slower machines/connections

kylo252 reviewed (commented):
almost there! 🔥

LostNeophyte commented:
I didn't test it on windows yet, when i do I'll check if the `/ -> \\` substitution is necessary and make sure everything works

LostNeophyte commented:
tested on windows and linux

kylo252 commented on utils/packaging/CMakeLists.txt.in:
```diff
@@ -37,17 +37,21 @@ message(STATUS "Using LVIM_VERSION: ${LVIM_VERSION}")
 else()
+  set(LVIM_INPUT_BIN_NAME lvim.template)
   set(LVIM_BIN_NAME lvim)
+  set(BASE_DIR_VAR "\$(readlink -f \"\$(dirname \"\$(realpath \"\$0\")\")/..\")/${CMAKE_INSTALL_DATAROOTDIR}/lunarvim")
```
how about using [`PROJECT_SOURCE_DIR`](https://cmake.org/cmake/help/latest/variable/PROJECT_SOURCE_DIR.html)? unless you're referencing the destination, then it should just be using `CMAKE_INSTALL_xxx`.

if it's easier, you could make it "build" the scripts and then you can install them according to installation rules.

LostNeophyte replied:
it's for the destination, it needs to be relative for the appimage, the paths are random (e.g /tmp/.mount_lvim(1mk5W7d/usr/share/lunarvim)

kylo252 replied:
> it needs to be relative for the appimage

do you mean _inside_ the appimage rather than relative?  What I'm trying to say is that having `/../` somewhere in path doesn't necessarily make it relative. I don't see how `BASE_DIR_VAR` wouldn't look like this  `/some/path/../share/lunarvim`, _which might even get resolved when using `readlink`_.

LostNeophyte replied:
we can't use a fixed absolute path https://docs.appimage.org/reference/best-practices.html?highlight=relocatable#binaries-must-not-use-compiled-in-absolute-paths
the suggested way to deal with this is to make paths relative to the executable, so here the path is `PATH_TO_EXE/../${CMAKE_INSTALL_DATAROOTDIR}/lunarvim` (`/tmp/.mount_lvim(1mk5W7d/usr/bin/../share/lunarvim` in appimage, `/usr/bin/../share/lunarvim` in a normal installation)

kylo252 commented on utils/installer/install.ps1:
```diff
@@ -184,7 +184,8 @@ function setup_shim() {
     }
 
-    Copy-Item -Force "$env:LUNARVIM_BASE_DIR\utils\bin\lvim.ps1" "$INSTALL_PREFIX\bin\lvim.ps1"
+    ((Get-Content -path "$env:LUNARVIM_BASE_DIR\utils\bin\lvim.ps1" -Raw) -replace '@BASE_DIR_VAR@','$env:LUNARVIM_RUNTIME_DIR\lvim') |
```
was this difficult to handle with cmake as well?

LostNeophyte replied:
I didn't think this one through, I also made this relative but appimage isn't for windows :facepalm:
